discord/file.py

```diff
@@ -100,18 +107,30 @@ def __init__(
 
         if filename is None:
             if isinstance(fp, str):
-                _, self.filename = os.path.split(fp)
+                _, filename = os.path.split(fp)
             else:
-                self.filename = getattr(fp, 'name', None)
-        else:
-            self.filename: Optional[str] = filename
+                filename = getattr(fp, 'name', 'untitled')
 
-        if spoiler and self.filename is not None and not self.filename.startswith('SPOILER_'):
-            self.filename = 'SPOILER_' + self.filename
+        self._filename, filename_spoiler = _strip_spoiler(filename)  # type: ignore # the above getattr doesn't narrow the type
+        if spoiler is MISSING:
+            spoiler = filename_spoiler
 
-        self.spoiler: bool = spoiler or (self.filename is not None and self.filename.startswith('SPOILER_'))
+        self.spoiler: bool = spoiler
         self.description: Optional[str] = description
 
+    @property
+    def filename(self) -> str:
+        """:class:`str`:
+        The filename to display when uploading to Discord.
+        If this is not given then it defaults to ``fp.name`` or if ``fp`` is
+        a string then the ``filename`` will default to the string given.
+        """
+        return 'SPOILER_' * self.spoiler + self._filename
```

Comment:
```suggestion
        return 'SPOILER_' + self._filename if self.spoiler else self._filename
```

Don't be too clever.